File: Sample/Sample/Settings View/SettingsViewController.m
/* Copyright Airship and Contributors */

@import AirshipCore;
@import AirshipLocation;

#import "SettingsViewController.h"


@interface SettingsViewController ()

@property (nonatomic, copy) NSString *localizedNone;
@property NSIndexPath* pushEnabled;
@property NSIndexPath* locationEnabled;
@property NSIndexPath* channelID;
@property NSIndexPath* namedUser;
@property NSIndexPath* tags;
@property NSIndexPath* analyticsEnabled;
@end

@implementation SettingsCell

- (void)layoutSubviews {
    [super layoutSubviews];
    
    NSString *sub = @"";
    
    if (sub == self.subtitle.text) {
        return;
    }
    
    if (!self.subtitle.text.length) {
        self.titleTopConstraint.priority = 100;
    } else {
        self.titleTopConstraint.priority = 999;
    }
}

@end

@implementation SettingsViewController

NSString *tagsSegue = @"tagsSegue";

static NSUInteger const sectionCount = 4;

typedef NS_ENUM(NSUInteger, UASettingsViewControllerSection) {
    UASettingsViewControllerSectionPush,
    UASettingsViewControllerSectionDevice,
    UASettingsViewControllerSectionAnalytics,
    UASettingsViewControllerSectionLocation
};

- (void)pushSettingsButtonTapped {
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString] options:@{} completionHandler:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIImage *image = [UIImage imageNamed:@"outline_settings_black_24pt"];
    
    UIBarButtonItem *pushSettings = [[UIBarButtonItem alloc]initWithImage:image style:UIBarButtonItemStyleDone target:self action:@selector(pushSettingsButtonTapped)];
    
    self.navigationItem.rightBarButtonItem = pushSettings;
    [self.navigationController.navigationBar setTitleTextAttributes: @{NSForegroundColorAttributeName:[UIColor whiteColor]}];

    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(refreshView) name:UAChannelUpdatedEvent object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didBecomeActive) name: UIApplicationDidBecomeActiveNotification object:nil];
    
    self.pushEnabled = [NSIndexPath indexPathForRow:0 inSection:0];
    self.channelID = [NSIndexPath indexPathForRow:0 inSection:1];
    self.namedUser = [NSIndexPath indexPathForRow:1 inSection:1];
    self.tags = [NSIndexPath indexPathForRow:2 inSection:1];
    self.locationEnabled = [NSIndexPath indexPathForRow:0 inSection:2];
    self.analyticsEnabled = [NSIndexPath indexPathForRow:0 inSection:3];
    self.localizedNone = NSLocalizedStringFromTable(@"ua_none", @"UAPushUI", @"None");
}

- (void)setTableViewTheme {
    self.tableView.backgroundColor = [UIColor whiteColor];
    [self.navigationController.navigationBar setTitleTextAttributes: @{NSForegroundColorAttributeName:[UIColor whiteColor]}];
    self.navigationController.navigationBar.barTintColor = [UIColor blackColor];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
}

- (void)didBecomeActive {
    [self refreshView];
    [self setTableViewTheme];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self refreshView];
}

- (void)refreshView {
    [self.tableView reloadData];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return sectionCount;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    switch (section) {
        case UASettingsViewControllerSectionPush:
            return  NSLocalizedStringFromTable(@"ua_device_info_push_settings", @"UAPushUI", @"Push Settings");
        case UASettingsViewControllerSectionDevice:
            return NSLocalizedStringFromTable(@"ua_device_info_device_settings", @"UAPushUI", @"Device Settings");
        case UASettingsViewControllerSectionAnalytics:
            return NSLocalizedStringFromTable(@"ua_device_info_analytics_settings", @"UAPushUI", @"Analytics Settings");
        case UASettingsViewControllerSectionLocation:
            return NSLocalizedStringFromTable(@"ua_device_info_location_settings", @"UAPushUI", @"Location Settings");
        default:
            return @"";
    }
}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)headerView forSection:(NSInteger)section {
    if([headerView isKindOfClass:[UITableViewHeaderFooterView class]]){
        UITableViewHeaderFooterView *tableViewHeaderFooterView = (UITableViewHeaderFooterView *) headerView;
        [tableViewHeaderFooterView.textLabel setTextColor:[UIColor colorWithRed:0.00 green:0.29 blue:1.00 alpha:1.00]];
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    switch (section) {
        case UASettingsViewControllerSectionPush:
            return 1;
        case UASettingsViewControllerSectionDevice:
            return 3;
        case UASettingsViewControllerSectionAnalytics:
            return 1;
        case UASettingsViewControllerSectionLocation:
            return 1;
        default:
            return 0;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"SettingsCell";
    
    SettingsCell *cell = [self.tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    cell.backgroundColor = [UIColor whiteColor];
    cell.title.textColor = [UIColor blackColor];
    cell.subtitle.textColor = [UIColor colorWithRed:0.51 green:0.51 blue:0.53 alpha:1.0];
    cell.cellSwitch.onTintColor = [UIColor colorWithRed:0.00 green:0.29 blue:1.00 alpha:1.00];
    cell.cellSwitch.tintColor = [UIColor colorWithRed:0.00 green:0.29 blue:1.00 alpha:1.00];
    
    cell.cellSwitch.hidden = YES;
    cell.cellSwitch.userInteractionEnabled = NO;
    cell.accessoryType = 0;
    cell.subtitle.text = nil;

    if (indexPath.section == UASettingsViewControllerSectionPush) {
        if (indexPath.row == self.pushEnabled.row) {
            cell.title.text = NSLocalizedStringFromTable(@"ua_device_info_push_settings", @"UAPushUI", @"Push Settings");
            cell.subtitle.text = NSLocalizedStringFromTable(@"ua_device_info_enable_push", @"UAPushUI", @"Enable Push");
            cell.cellSwitch.hidden = NO;
            cell.cellSwitch.on = UAirship.push.userPushNotificationsEnabled;
            cell.subtitle.text = [self pushTypeString];
            cell.subtitle.adjustsFontSizeToFitWidth = YES;
            cell.subtitle.minimumScaleFactor = 0.25;
            cell.subtitle.numberOfLines = 1;
        }
    }
    if (indexPath.section == UASettingsViewControllerSectionDevice) {
        if (indexPath.row == self.channelID.row) {
            cell.title.text = NSLocalizedStringFromTable(@"ua_device_info_channel_id", @"UAPushUI", @"Channel ID");
            cell.subtitle.text = UAirship.channel.identifier;
        }
    }
    if (indexPath.row == self.namedUser.row) {
        cell.title.text = NSLocalizedStringFromTable(@"ua_device_info_named_user", @"UAPushUI", @"Named User");
         cell.subtitle.text = (UAirship.namedUser.identifier == nil) ? self.localizedNone : UAirship.namedUser.identifier;
         cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    if (indexPath.row == self.tags.row) {
         cell.title.text = NSLocalizedStringFromTable(@"ua_device_info_tags", @"UAPushUI", @"Tags");
         if (UAirship.channel.tags.count > 0) {
             cell.subtitle.text = [UAirship.channel.tags componentsJoinedByString:@", "];
         } else {
             cell.subtitle.text = self.localizedNone;
         }

         cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    if (indexPath.section == UASettingsViewControllerSectionAnalytics) {
        if (indexPath.row == self.analyticsEnabled.row) {
            cell.title.text = NSLocalizedStringFromTable(@"ua_device_info_analytics_enabled", @"UAPushUI", @"Analytics Enabled");
            cell.subtitle.text = NSLocalizedStringFromTable(@"ua_device_info_enable_analytics_tracking", @"UAPushUI", @"Enable analytics tracking");
            cell.cellSwitch.hidden = NO;
            cell.cellSwitch.on = UAirship.analytics.isEnabled;
            cell.subtitle.adjustsFontSizeToFitWidth = YES;
        }
    }
    if (indexPath.section == UASettingsViewControllerSectionLocation) {
        if (indexPath.row == self.locationEnabled.row) {
            cell.title.text = NSLocalizedStringFromTable(@"ua_device_info_enable_location_enabled", @"UAPushUI", @"Location Enabled");
            cell.cellSwitch.hidden = NO;
            cell.cellSwitch.on = UALocation.shared.isLocationUpdatesEnabled;
            BOOL optedInToLocation = UALocation.shared.isLocationOptedIn;
            if(UALocation.shared.isLocationUpdatesEnabled && !optedInToLocation) {
                cell.subtitle.text = [NSLocalizedStringFromTable(@"ua_location_enabled_detail", @"UAPushUI", @"Enable GPS and WIFI Based Location detail label") stringByAppendingString:@" - NOT OPTED IN"];
            } else {
                cell.subtitle.text = self.localizedNone;
            }
        }
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    SettingsCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
    
    if (indexPath.section == 0) {
        if (indexPath.row == self.pushEnabled.row) {
            [cell.cellSwitch setOn:!cell.cellSwitch.on animated:YES];
            UAirship.push.userPushNotificationsEnabled = cell.cellSwitch.on;
        }
    }
    if (indexPath.section == 1) {
        if (indexPath.row == self.channelID.row) {
            if (UAirship.channel.identifier != nil) {
                UIPasteboard.generalPasteboard.string = cell.subtitle.text;
                [self showCopiedAlert];
            }
        }
        if (indexPath.row == self.namedUser.row) {
            [self performSegueWithIdentifier:@"namedUserSegue" sender:self];
        }
        if (indexPath.row == self.tags.row) {
            [self performSegueWithIdentifier:@"tagsSegue" sender:self];

        }
    }
    if (indexPath.section == 2) {
        if (indexPath.row == self.analyticsEnabled.row) {
            [cell.cellSwitch setOn:!cell.cellSwitch.on animated:YES];
            UAirship.push.userPushNotificationsEnabled = cell.cellSwitch.on;
        }
    }
    if (indexPath.section == 3) {
        if (indexPath.row == self.locationEnabled.row) {
            [cell.cellSwitch setOn:!cell.cellSwitch.on animated:YES];
            UAirship.push.userPushNotificationsEnabled = cell.cellSwitch.on;
        }
    }

    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void)showCopiedAlert {
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:nil
                                                                       message:NSLocalizedStringFromTable(@"UA_Copied_To_Clipboard", @"UAPushUI", @"Copied to clipboard string")
                                                                preferredStyle:UIAlertControllerStyleAlert];

        UIAlertAction *okAction = [UIAlertAction actionWithTitle:NSLocalizedStringFromTable(@"UA_OK", @"UAPushUI", @"OK button string")
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * _Nonnull action) {
                                                             [self dismissViewControllerAnimated:YES completion:nil];
                                                         }];

        [alert addAction:okAction];

        [self presentViewController:alert animated:YES completion:nil];
}

- (NSString *)pushTypeString {
    UAAuthorizedNotificationSettings authorizedSettings = [UAirship push].authorizedNotificationSettings;

    
    NSMutableArray *settingsArray = [[NSMutableArray alloc] init];;
    
    if ((UAAuthorizedNotificationSettingsAlert & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_alerts", @"UAPushUI", @"Alerts")];
    }
    
    if ((UAAuthorizedNotificationSettingsBadge & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_badges", @"UAPushUI", @"Badges")];
    }
    
    if ((UAAuthorizedNotificationSettingsSound & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_sounds", @"UAPushUI", @"Sounds")];
    }
    
    if ((UAAuthorizedNotificationSettingsCarPlay & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_car_play", @"UAPushUI", @"CarPlay")];
    }

    if ((UAAuthorizedNotificationSettingsLockScreen & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_lock_screen", @"UAPushUI", @"Lock Screen")];
    }

    if ((UAAuthorizedNotificationSettingsNotificationCenter & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_notification_center", @"UAPushUI", @"Notification Center")];
    }
    
    if ((UAAuthorizedNotificationSettingsCriticalAlert & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_critical_alert", @"UAPushUI", @"Critical Alert")];
    }
    
    if ((UAAuthorizedNotificationSettingsAnnouncement & authorizedSettings) > 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_notification_type_announcement", @"UAPushUI", @"AirPod Announcement")];
    }
    
    if (settingsArray.count == 0) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"ua_push_settings_link_disabled_title", @"UAPushUI", @"Pushes Currently Disabled")];
    }
    
    return [settingsArray componentsJoinedByString:(@", ")];
}

@end

File: SwiftSample/SwiftSample/SettingsViewController.swift
/* Copyright Airship and Contributors */

import UIKit


#if canImport(AirshipCore)
import AirshipCore
import AirshipAutomation
import AirshipMessageCenter
import AirshipLocation
#elseif canImport(Airship)
import Airship
#endif


class SettingsCell: UITableViewCell {
    @IBOutlet weak var title: UILabel!
    @IBOutlet weak var subtitle: UILabel!
    @IBOutlet weak var cellSwitch: UISwitch!

    @IBOutlet var titleTopConstraint: NSLayoutConstraint!

    override func layoutSubviews() {
        super.layoutSubviews()

        guard let sub = subtitle.text else {
            return
        }

        if sub.isEmpty {
            titleTopConstraint.priority = UILayoutPriority(rawValue: 100)
        } else {
            titleTopConstraint.priority = UILayoutPriority(rawValue: 999)
        }

        layoutIfNeeded()
    }
    
}

class SettingsViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    @objc public static var tagsViewName = "tags"
    let tagsSegue = "tagsSegue"

    var launchPathComponents : [String]? {
        didSet {
            if (self.viewIfLoaded?.window != nil && self === self.navigationController?.visibleViewController) {
                segueToDeepLink();
            } else if (!(self  === self.navigationController?.visibleViewController))  {
                self.navigationController?.popToRootViewController(animated: false);
            }
        }
    }
    var launchCompletionHandler : (() -> Void)?

    private let localizedNone =  "ua_none".localized(comment: "None")
    private let sectionCount = 4
    
    @IBOutlet private var tableView: UITableView!

    /* Section
     * Note: Number of sections and sections for row are defined in their respective
     * table view data source methods
     */
    let pushSettings = 0,
    deviceSettings = 1,
    locationSettings = 2,
    analyticsSettings = 3

    // Push settings
    private let pushEnabled = IndexPath(row: 0, section: 0)
    
    // Device settings
    private let channelID = IndexPath(row: 0, section: 1),
    namedUser = IndexPath(row: 1, section: 1),
    tags = IndexPath(row: 2, section: 1)

    // Location settings
    private let locationEnabled = IndexPath(row: 0, section: 2)

    // Analytics settigns
    private let analyticsEnabled = IndexPath(row: 0, section: 3)
    
    @objc func pushSettingsButtonTapped(sender:Any) {
        UIApplication.shared.open(URL(string: UIApplication.openSettingsURLString)!, options: [:],completionHandler: nil)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let image = UIImage(named: "outline_settings_black_24pt")

        let pushSettings: UIBarButtonItem = UIBarButtonItem(image:image, style:.done, target: self, action: #selector(pushSettingsButtonTapped))
        self.navigationItem.rightBarButtonItem = pushSettings

        tableView.dataSource = self
        tableView.delegate = self
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(SettingsViewController.refreshView),
            name: Channel.channelCreatedEvent,
            object: nil);
        
        // add observer to didBecomeActive to update upon retrun from system settings screen
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(SettingsViewController.didBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil)
    }

    func setTableViewTheme() {
        tableView.backgroundColor = #colorLiteral(red: 0.999904573, green: 1, blue: 0.9998808503, alpha: 1)
        navigationController?.navigationBar.titleTextAttributes = [.foregroundColor:#colorLiteral(red: 0.999904573, green: 1, blue: 0.9998808503, alpha: 1)]
        navigationController?.navigationBar.barTintColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 1);
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated);
        segueToDeepLink()
    }
    
    func segueToDeepLink() {
        if let launchPathComponents = launchPathComponents {
            self.launchPathComponents = nil
            if (launchPathComponents.count > 0) {
                switch (launchPathComponents[0]) {
                case "tags":
                    performSegue(withIdentifier: tagsSegue, sender: self)
                default:
                    break
                }
            }
        }
        if let launchCompletionHandler = launchCompletionHandler {
            launchCompletionHandler()
        }
    }
    
    // this is necessary to update the view when returning from the system settings screen
    @objc func didBecomeActive() {
        refreshView()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        refreshView()
        setTableViewTheme()
    }
    
    @objc func refreshView() {
        tableView.reloadData()
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return sectionCount
    }

    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case pushSettings:
            return "ua_device_info_push_settings".localized(comment: "Push Settings")
        case deviceSettings:
            return "ua_device_info_device_settings".localized(comment: "Device Settings")
        case analyticsSettings:
            return "ua_device_info_analytics_settings".localized(comment: "Analytics")
        case locationSettings:
            return "ua_device_info_enable_location_enabled".localized(comment: "Location Enabled")
        default:
            return ""
        }
    }

    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        if let headerView = view as? UITableViewHeaderFooterView {
            headerView.textLabel?.textColor = #colorLiteral(red: 0, green: 0.2950756848, blue: 0.9987069964, alpha: 1)
        }
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case pushSettings:
            return 1
        case deviceSettings:
            return 3
        case analyticsSettings:
            return 1
        case locationSettings:
            return 1
        default:
            return 0
        }
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell:SettingsCell = tableView.dequeueReusableCell(withIdentifier: "SettingsCell", for: indexPath) as! SettingsCell

        cell.backgroundColor = #colorLiteral(red: 0.999904573, green: 1, blue: 0.9998808503, alpha: 1)
        cell.title.textColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 1)
        cell.subtitle?.textColor = #colorLiteral(red: 0.513617754, green: 0.5134617686, blue: 0.529979229, alpha: 1)
        cell.cellSwitch.onTintColor = #colorLiteral(red: 0, green: 0.2941176471, blue: 1, alpha: 1)
        cell.cellSwitch.tintColor = #colorLiteral(red: 0, green: 0.2950756848, blue: 0.9987069964, alpha: 1)

        // Cell switch, stepper and disclosure indicator are hidden by default
        cell.cellSwitch.isHidden = true

        // Cells will do the switching
        cell.cellSwitch.isUserInteractionEnabled = false
        cell.accessoryType = .none
        cell.subtitle.text = nil

        switch indexPath {
        case pushEnabled:
            cell.title.text = "ua_device_info_push_settings".localized(comment: "Push Settings")
            cell.subtitle.text = "ua_device_info_enable_push".localized(comment: "Enable Push")
            cell.cellSwitch.isHidden = false
            cell.cellSwitch.isOn = UAirship.push()?.userPushNotificationsEnabled ?? false
            cell.subtitle?.text = pushTypeString()
            cell.subtitle?.adjustsFontSizeToFitWidth = true;
            cell.subtitle?.minimumScaleFactor = 0.25;
            cell.subtitle?.numberOfLines = 1;
        case channelID:
            cell.title.text = "ua_device_info_channel_id".localized(comment: "Channel ID")
            cell.subtitle.text = UAirship.channel().identifier
        case namedUser:
            cell.title.text = "ua_device_info_named_user".localized(comment: "Named User")
            cell.subtitle?.text = UAirship.namedUser().identifier == nil ? localizedNone : UAirship.namedUser().identifier
            cell.accessoryType = .disclosureIndicator
        case tags:
            cell.title.text = "ua_device_info_tags".localized(comment: "Tags")
            if (UAirship.channel().tags.count > 0) {
                cell.subtitle?.text = UAirship.channel().tags.joined(separator: ", ")
            } else {
                cell.subtitle?.text = localizedNone
            }

            cell.accessoryType = .disclosureIndicator
        case analyticsEnabled:
            cell.title.text = "ua_device_info_analytics_enabled".localized(comment: "Analytics Enabled")
            cell.subtitle?.text = "ua_device_info_enable_analytics_tracking".localized(comment: "Enable analytics tracking")

            cell.cellSwitch.isHidden = false
            cell.cellSwitch.isOn = UAirship.shared().privacyManager.isEnabled(UAFeatures.analytics)
        case locationEnabled:
            cell.title.text = "ua_device_info_location_settings".localized(comment: "Location Settings")

            cell.cellSwitch.isHidden = false
            cell.cellSwitch.isOn = UAirship.shared().privacyManager.isEnabled(UAFeatures.location)

            let optedInToLocation = UALocation.shared().isLocationOptedIn()

            if (UAirship.shared().privacyManager.isEnabled(UAFeatures.location) && !optedInToLocation) {
                cell.subtitle?.text = "ua_location_enabled_detail".localized(comment: "Enable GPS and WIFI Based Location detail label") + " - NOT OPTED IN"
            } else {
                cell.subtitle?.text = localizedNone
            }
        default:
            break
        }

        return cell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let cell = tableView.cellForRow(at: indexPath) as! SettingsCell

        switch indexPath {
        case pushEnabled:
            cell.cellSwitch.setOn(!cell.cellSwitch.isOn, animated: true)
            if (cell.cellSwitch.isOn) {
                UAirship.push().userPushNotificationsEnabled = true
                UAirship.shared().privacyManager.enableFeatures(.push)
            } else {
                UAirship.push().userPushNotificationsEnabled = false
                UAirship.shared().privacyManager.disableFeatures(.push)
            }

        case channelID:
            if (UAirship.channel().identifier != nil) {
                UIPasteboard.general.string = cell.subtitle?.text
                showCopiedAlert()
            }
        case namedUser:
            performSegue(withIdentifier: "namedUserSegue", sender: self)
        case tags:
            performSegue(withIdentifier: "tagsSegue", sender: self)
        case analyticsEnabled:
            cell.cellSwitch.setOn(!cell.cellSwitch.isOn, animated: true)
            if (cell.cellSwitch.isOn) {
                UAirship.shared().privacyManager.enableFeatures(UAFeatures.analytics)
            } else {
                UAirship.shared().privacyManager.disableFeatures(UAFeatures.analytics)
            }
        case locationEnabled:
            cell.cellSwitch.setOn(!cell.cellSwitch.isOn, animated: true)
            if (cell.cellSwitch.isOn) {
                UAirship.shared().privacyManager.enableFeatures(UAFeatures.location)
                UALocation.shared().isLocationUpdatesEnabled = true
            } else {
                UAirship.shared().privacyManager.disableFeatures(UAFeatures.location)
                UALocation.shared().isLocationUpdatesEnabled = false
            }
        default:
            break
        }
        
        tableView.deselectRow(at: indexPath, animated: true)
    }

    func showCopiedAlert() {
        DispatchQueue.main.async {
            let message = "ua_copied_to_clipboard".localized(comment: "Copied to clipboard string")

            let alert = UIAlertController(title: nil, message: message, preferredStyle: UIAlertController.Style.alert)
            let buttonTitle = "UA_OK".localized(comment: "OK")

            let okAction = UIAlertAction(title: buttonTitle, style: UIAlertAction.Style.default, handler: { (action: UIAlertAction) in
                self.dismiss(animated: true, completion: nil)
            })

            alert.addAction(okAction)

            self.present(alert, animated: true, completion: nil)
        }
    }

    func pushTypeString () -> String {

        let authorizedSettings = UAirship.push().authorizedNotificationSettings;

        var settingsArray: [String] = []

        if (authorizedSettings.contains(.alert)) {
            settingsArray.append("ua_notification_type_alerts".localized(comment: "Alerts"))
        }

        if (authorizedSettings.contains(.badge)){
            settingsArray.append("ua_notification_type_badges".localized(comment: "Badges"))
        }

        if (authorizedSettings.contains(.sound)) {
            settingsArray.append("ua_notification_type_sounds".localized(comment: "Sounds"))
        }

        if (authorizedSettings.contains(.carPlay)) {
            settingsArray.append("ua_notification_type_car_play".localized(comment: "CarPlay"))
        }

        if (authorizedSettings.contains(.lockScreen)) {
            settingsArray.append("ua_notification_type_lock_screen".localized(comment: "Lock Screen"))
        }

        if (authorizedSettings.contains(.notificationCenter)) {
            settingsArray.append("ua_notification_type_notification_center".localized(comment: "Notification Center"))
        }

        if (authorizedSettings.contains(.criticalAlert)) {
            settingsArray.append("ua_notification_type_critical_alert".localized(comment: "Critical Alert"))
        }

        if (authorizedSettings.contains(.announcement)) {
            settingsArray.append("ua_notification_type_announcement".localized(comment: "AirPod Announcement"))
        }
        
        if (settingsArray.count == 0) {
            settingsArray.append("ua_push_settings_link_disabled_title".localized(comment: "Pushes Currently Disabled"))
        }

        return settingsArray.joined(separator: ", ")
    }
}

internal extension String {
    func localized(tableName: String = "UAPushUI", comment: String = "") -> String {
        return NSLocalizedString(self, tableName: tableName, comment: comment)
    }
    
    func localizedWithFormat(count:Int) -> String {
        return String.localizedStringWithFormat(localized(), count)
    }
}
